agent/src/util/discovery_configuration_controller.rs

```diff
@@ -0,0 +1,638 @@
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex},
+    time::Duration,
+};
+
+use akri_shared::{
+    akri::{
+        configuration::{Configuration, DiscoveryProperty},
+        instance::Instance,
+    },
+    k8s::crud::IntoApi,
+};
+use futures::StreamExt;
+use tokio::sync::mpsc;
+
+use crate::discovery_handler_manager::{
+    discovery_handler_registry::DiscoveryHandlerRegistry, DiscoveryError,
+};
+
+use kube::{Resource, ResourceExt};
+use kube_runtime::{
+    controller::Action,
+    reflector::{ObjectRef, Store},
+    Controller,
+};
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error(transparent)]
+    DiscoveryError(#[from] DiscoveryError),
+    #[error(transparent)]
+    Other(#[from] anyhow::Error),
+}
+
+pub trait DiscoveryConfigurationKubeClient: IntoApi<Configuration> + IntoApi<Instance> {}
+
+impl<T: IntoApi<Configuration> + IntoApi<Instance>> DiscoveryConfigurationKubeClient for T {}
+
+pub struct ControllerContext {
+    pub instances_cache: Store<Instance>,
+    pub dh_registry: Arc<dyn DiscoveryHandlerRegistry>,
+    pub client: Arc<dyn DiscoveryConfigurationKubeClient>,
+    pub agent_instance_name: String,
+    pub error_backoffs: Mutex<HashMap<String, Duration>>,
+}
+
+pub async fn start_controller(
+    ctx: Arc<ControllerContext>,
+    rec: mpsc::Receiver<ObjectRef<Configuration>>,
+) {
+    let api = ctx.client.all().as_inner();
+    let controller = Controller::new(api, Default::default());
+
+    controller
+        .graceful_shutdown_on(async {
+            let mut signal =
+                tokio::signal::unix::signal(tokio::signal::unix::SignalKind::terminate()).unwrap();
+            signal.recv().await;
+        })
+        .reconcile_on(tokio_stream::wrappers::ReceiverStream::new(rec))
+        .run(reconcile, error_policy, ctx)
+        .for_each(|_| futures::future::ready(()))
+        .await;
+}
+
+pub async fn reconcile(
+    dc: Arc<Configuration>,
+    ctx: Arc<ControllerContext>,
+) -> Result<Action, Error> {
+    trace!("Reconciling {:?}::{}", dc.namespace(), dc.name_any());
+    let namespace = dc.namespace().unwrap();
+    let owner_ref = dc.controller_owner_ref(&()).unwrap();
+    if dc.metadata.deletion_timestamp.is_some() {
+        ctx.dh_registry.terminate_request(&dc.name_any()).await;
+
+        ctx.client
+            .namespaced(&namespace)
+            .remove_finalizer(dc.as_ref(), &ctx.agent_instance_name)
+            .await
+            .map_err(|e| Error::Other(e.into()))?;
+
+        return Ok(Action::await_change());
+    }
+
+    if !dc.finalizers().contains(&ctx.agent_instance_name) {
```

Comment:
Is this essentially a start up action for the reconciler? Adds the finalizers if it hasn't been added yet? Or is this case met in other scenarios?